pkg/infrastructure/azure/azure.go

```diff
@@ -105,9 +112,117 @@ func (p *Provider) PreProvision(ctx context.Context, in clusterapi.PreProvisionI
 	if err != nil {
 		return fmt.Errorf("error creating resource group %s: %w", resourceGroupName, err)
 	}
+	resourceGroup, err := resourceGroupsClient.Get(ctx, resourceGroupName, nil)
+	if err != nil {
+		return fmt.Errorf("error getting resource group %s: %w", resourceGroupName, err)
+	}
+
 	logrus.Debugf("ResourceGroup.ID=%s", *resourceGroup.ID)
 	p.ResourceGroupName = resourceGroupName
 
+	// Create user assigned identity
+	userAssignedIdentityName := fmt.Sprintf("%s-identity", in.InfraID)
+	armmsiClientFactory, err := armmsi.NewClientFactory(
+		subscriptionID,
+		tokenCredential,
+		&arm.ClientOptions{
+			ClientOptions: policy.ClientOptions{
+				Cloud: cloudConfiguration,
+			},
+		},
+	)
+	if err != nil {
+		return fmt.Errorf("failed to create armmsi client: %w", err)
+	}
+	_, err = armmsiClientFactory.NewUserAssignedIdentitiesClient().CreateOrUpdate(
+		ctx,
+		resourceGroupName,
+		userAssignedIdentityName,
+		armmsi.Identity{
+			Location: ptr.To(platform.Region),
+			Tags:     tags,
+		},
+		nil,
+	)
+	if err != nil {
+		return fmt.Errorf("failed to create user assigned identity %s: %w", userAssignedIdentityName, err)
+	}
+	userAssignedIdentity, err := armmsiClientFactory.NewUserAssignedIdentitiesClient().Get(
+		ctx,
+		resourceGroupName,
+		userAssignedIdentityName,
+		nil,
+	)
+	if err != nil {
+		return fmt.Errorf("failed to get user assigned identity %s: %w", userAssignedIdentityName, err)
+	}
+	principalID := *userAssignedIdentity.Properties.PrincipalID
+
+	logrus.Debugf("UserAssignedIdentity.ID=%s", *userAssignedIdentity.ID)
+	logrus.Debugf("PrinciapalID=%s", principalID)
+
+	clientFactory, err := armauthorization.NewClientFactory(
+		subscriptionID,
+		tokenCredential,
+		&arm.ClientOptions{
+			ClientOptions: policy.ClientOptions{
+				Cloud: cloudConfiguration,
+			},
+		},
+	)
+	if err != nil {
+		return fmt.Errorf("failed to create armauthorization client: %w", err)
+	}
+
+	roleDefinitionsClient := clientFactory.NewRoleDefinitionsClient()
+
+	var contributor *armauthorization.RoleDefinition
+	roleDefinitionsPager := roleDefinitionsClient.NewListPager(*resourceGroup.ID, nil)
```

Comment:
It's hard for me to tell exactly what `Filter` in the `RoleDefinitionsClientListOptions` does, but if we can include a filter to narrow down the list that would be good. I can imagine these Role Definition Lists may be large:
https://beta.pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/authorization/armauthorization/v3#RoleDefinitionsClientListOptions

Reply:
My interpretation of the filter was scope based. Since there is absolutely no documentation on how to use it, I've left it alone.

Reply:
This is the only information I found on scoping filters: https://learn.microsoft.com/en-us/entra/identity/app-provisioning/define-conditional-rules-for-provisioning-user-accounts?pivots=app-provisioning&WT.mc_id=AZ-MVP-5003178

Not sure if that really helps.

Reply:
> My interpretation of the filter was scope based. 

Yeah it seems `atScopeAndBelow`, whatever that means, is the only supported filter. 

Reviewing [the API reference](https://learn.microsoft.com/en-us/rest/api/authorization/role-definitions/list?view=rest-authorization-2022-04-01&tabs=HTTP), I can't see any way to limit this.

Hopefully this won't become a burden in situations where users may have lots of custom roles. At least `Contributor` is relatively early in the alphabet..